readme.md
-------

# UNANSWERED TOPICS v2.7

[**By Dougiefresh**](http://www.simplemachines.org/community/index.php?action=profile;u=253913) -> [Link to Mod](http://custom.simplemachines.org/mods/index.php?mod=4088)

-------

## Introduction
This mod shows all of the topics, excluding locked topics and recycle board topics, that have not been answered (aka topics that consist of a single post).  Locked topics and topics in the recycle boards are not included in the list because it makes no sense to include topics that can't be replied to....

**Version 2.5** adds the ability to restrict by board(s) by using **board=[x]**, where **[x]** is the board ID or IDs seperated by commas.

## Admin Settings
There are two new settings under **Admin** -> **Configuration** -> **Modification Settings**:

- Default view for unanswered topics (set to posts view by default)
- Number of days to show unanswered topics (set to 90 days by default)

There is also a new permission under **Admin** -> **Members** -> **Permissions**:

- Can view unanswered topics (not permissable for guests)

## Compatibility Notes
This mod was tested on SMF 2.0.13 and SMF 2.1 Beta 2, but should work on SMF 2.0 and up.  SMF 1.x is not and will not be supported.

## Related Discussions

- [Un-answeredTopics for SMF 2.0?](http://www.simplemachines.org/community/index.php?topic=510124.0[/url)

## Changelog
The changelog can be viewed at [XPtsp.com](http://www.xptsp.com/board/free-modifications/unanswered-topics/?tab=1).

## License
Copyright (c) 2016 - 2018, Douglas Orend

All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changelog.txt
v2.7 - November 9th, 2018
o Updated Spanish Latin translation, courtsey of [url=https://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee[/url]!

v2.6 - November 9th, 2018
o No functionality change.
o Updated documentation to point to new website.

v2.5 - August 6th, 2017
o Added ability to narrow topic list by board ID.

v2.4 - July 28th, 2017
o Added permission to view unanswered topics.

v2.3 - July 11th, 2017
o Replaced IFNULL with COALESCE in database queries for better Postgres support.

v2.2 - July 5th, 2017
o SMF 2.1: Replaced missing [b]settings_install.php[/b] with [b]add_remove_hooks.php[/b] in [b]package-info.xml[/b].

v2.1 - January 7th, 2017
o Fixed text display issue when number of days is equal to 0 in several places.
o Removed unused topic/post option leftover from version 1.x series.
o Changed several language strings to reflect better what the mod does when days equal 0.

v2.0.1 - January 6th, 2017
o Fixed database query involving invalid field for this mod.

v2.0 - January 5th, 2017
o Converted mod to 100% hook-based, removing all file edits!
o Completely rewrote unanswered topics function to use built-in list functionality.
o Moved all language strings to individual language files.
o Added code to exclude contents of the recycle board from the list.
o Added checkboxes to remove topics, when appropriate permissions are granted to user.
o Added code to restrict unanswered topics to selected member, if desirable.

v1.7 - October 1st, 2016
o Fixed a bug where posts where displayed instead of topics on page 2 and up...  Thanks, [url=http://www.simplemachines.org/community/index.php?action=profile;u=413177]aegersz[/url]!

v1.6 - September 30th, 2016
o Fixed German language file so that it installs properly without errors.

v1.5 - Match 30th, 2016
o Added Spanish-Latin translation by [url=http://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee [BC][/url].

v1.4 - March 27th, 2016
o Second attempt to fix array processing issue found by [url=http://www.simplemachines.org/community/index.php?action=profile;u=189089]Inti31[/url].

v1.3 - March 19th, 2016
o Fixed "Array to string conversion" error found by [url=http://www.simplemachines.org/community/index.php?action=profile;u=423124]roza[/url].
o Fixed array processing issue found by [url=http://www.simplemachines.org/community/index.php?action=profile;u=189089]Inti31[/url].
o Added German translation by [url=http://www.simplemachines.org/community/index.php?action=profile;u=189089]Inti31[/url].

v1.2 - March 7th, 2016
o Added Turkish language support to the mod, thanks to Snrj!
o Added SMF 2.1 Beta 2 support to the mod.
o Moved 2 language strings from [b]index[/b] to [b]ManageSettings[/b] language file.

v1.1 - February 29th, 2016
o Added option to set number of days for unanswered topics under [b]Modification Settings[/b].
o Added option to set default view of the unanswered topics (topic or post).
o Added code to show unanswered topics as a topic list.
o Added buttons to switch between the two views: topic and post.
o Removed code that stops users from limiting search to boards and/or categories.
o Enabled all buttons to be shown (permissions permitting) when viewing the posts.

v1.0 - February 27th, 2016
o Initial release to the public

Unanswered.php
<?php
/********************************************************************************
* Unanswered.php - Subs of the Unanswered Topics mod
*********************************************************************************
* This program is distributed in the hope that it is and will be useful, but
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY
* or FITNESS FOR A PARTICULAR PURPOSE,
**********************************************************************************/
if (!defined('SMF'))
	die('Hacking attempt...');

/********************************************************************************
* Functions necessary to list & mark our "important topics" to the user:
********************************************************************************/
function UnansweredTopics()
{
	global $context, $txt, $scripturl, $modSettings, $smcFunc, $sourcedir;

	// Set up for listing the unanswered topics:
	loadLanguage('Unanswered');
	$context['page_title' ] = $txt['unanswered_topics'];
	$context['sub_template'] = 'unanswered_topics';
	$context['delete_own'] = $context['delete_any'] = array();
	$context['can_delete'] = false;
	$context['unanswered_member'] = isset($_GET['u']) ? (int) $_GET['u'] : false;
	$context['days'] = isset($modSettings['unanswered_time_limit']) ? $modSettings['unanswered_time_limit'] : 90;
	$context['UAT_boards'] = isset($_GET['board']) ? explode(',', $_GET['board']) : array();

	// Are we removing the topic?
	if (isset($_GET['save']) && isset($_POST['remove_submit']))
	{
		// Remove every topic we have permission to:
		checkSession();
		if (!empty($_POST['remove']))
		{
			require_once($sourcedir . '/RemoveTopic.php');
			$context['unanswered_topics'] = $topics = array();
			foreach ($_POST['remove'] AS $id_topic => $ignored)
			{
				if (in_array($id_topic, $_SESSION['unanswered_can_delete']))
					$topics[] = $id_topic;
				else
					$context['unanswered_topics'][] = $id_topic;
			}
			removeTopics($topics);
		}

		// If there are topics left, notify user about no permission to delete:
		if (empty($context['unanswered_topics']))
			redirectExit('action=unanswered;sort=' . $_POST['sort'] . (!empty($_POST['desc']) ? ';desc' : '') . (!empty($_POST['start']) ? ';start=' . $_POST['start'] : ''));
		$context['sub_template'] = 'unanswered_cannot_delete';
		return;
	}
	$_SESSION['unanswered_can_delete'] = array();

	// Set the options for the list component.
	$topic_listOptions = array(
		'id' => 'unanswered_topics',
		'title' => $txt['unanswered_topics'],
		'items_per_page' => $modSettings['defaultMaxMessages'],
		'base_href' => $scripturl . '?action=unanswered',
		'default_sort_col' => 'lastpost',
		'default_sort_dir' => 'desc',
		'no_items_label' => !empty($context['days']) ? sprintf($txt['unanswered_topics_visit_none'], $context['days']) : $txt['unanswered_topics_visit_unlimited'],
		'get_items' => array(
			'function' => 'UTM_Get_Topics',
		),
		'get_count' => array(
			'function' => 'UTM_Topics_Count',
		),
		'columns' => array(
			'icon' => array(
				'header' => array(
					'value' => '',
				),
				'data' => array(
					'function' => 'UTM_Icon',
					'style' => 'text-align: center; width: 30px',
				),
			),
			'subject' => array(
				'header' => array(
					'value' => $txt['topics'],
				),
				'data' => array(
					'function' => 'UTM_Subject',
				),
				'sort' => array(
					'default' => 'b.name, m.subject',
					'reverse' => 'b.name DESC, m.subject DESC',
				),
			),
			'views' => array(
				'header' => array(
					'value' => $txt['views'],
				),
				'data' => array(
					'function' => create_function('$rowData', '
						return comma_format($rowData["num_views"]);
					'),
					'style' => 'text-align: center; width: 7%',
				),
				'sort' => array(
					'default' => 't.num_views',
					'reverse' => 't.num_views DESC',
				),
			),
			'lastpost' => array(
				'header' => array(
					'value' => $txt['last_post'],
				),
				'data' => array(
					'function' => create_function('$rowData', '
						return timeformat($rowData["posted"]);
					'),
					'style' => 'width: 30%',
				),
				'sort' => array(
					'default' => 'm.poster_time',
					'reverse' => 'm.poster_time DESC',
				),
			),
			'check' => array(
				'header' => array(
					'value' => '<input type="checkbox" onclick="invertAll(this, this.form);" class="input_check" />',
				),
				'data' => array(
					'function' => 'UTM_Checkbox',
					'style' => 'text-align: center; width: 30px',
				),
			),
		),
		'form' => array(
			'href' => $scripturl . '?action=unanswered;save',
			'include_sort' => true,
			'include_start' => true,
		),
		'additional_rows' => array(
			array(
				'position' => 'below_table_data',
				'value' => '<input type="submit" name="remove_submit" class="button_submit" value="' . $txt['quick_mod_remove'] . '" onclick="return confirm(\'' . $txt['unanswered_remove_topics'] . '\');" />',
				'style' => 'text-align: right;',
			),
		),
	);

	// Create the list.
	require_once($sourcedir . '/Subs-List.php');
	createList($topic_listOptions);
	
	// Can we delete ANY threads in the list?  If not, remove the checkboxes!!!
	if (empty($context['can_delete']))
	{
		foreach ($context['unanswered_topics']['headers'] as $num => $element)
		{
			if ($element['id'] == 'check')
				unset($context['unanswered_topics']['headers'][$num]);
		}
		foreach ($context['unanswered_topics']['rows'] as $num => $element)
			unset($context['unanswered_topics']['rows'][$num]['check']);
		unset($context['unanswered_topics']['additional_rows']);
	}
}

/********************************************************************************
* Functions doing our database queries:
********************************************************************************/
function UTM_Topics_Count()
{
	global $smcFunc, $modSettings, $context;

	$request = $smcFunc['db_query']('', '
		SELECT COUNT(t.id_topic) AS count
		FROM {db_prefix}topics AS t
			LEFT JOIN {db_prefix}boards AS b ON (b.id_board = t.id_board)
			LEFT JOIN {db_prefix}messages AS m ON (m.id_msg = t.id_first_msg)
		WHERE {query_see_board}' . (!empty($modSettings['recycle_enable']) && !empty($modSettings['recycle_board']) ? '
			AND b.id_board != {int:recycle_board}' : '') . (!empty($context['unanswered_member']) ? '
			AND m.id_member = {int:id_member}' : '') . '
			AND t.id_first_msg = t.id_last_msg
			AND t.locked = {int:not_locked}' . (!empty($context['days']) ? '
			AND m.poster_time > {int:time_limit}' : '') . (!empty($context['UAT_boards']) ? '
			AND b.id_board IN ({array_int:boards})' : ''),
		array(
			'not_locked' => 0,
			'id_member' => $context['unanswered_member'],
			'time_limit' => time() - $context['days'] * 86400,
			'recycle_board' => $modSettings['recycle_board'],
			'boards' => $context['UAT_boards'],
		)
	);
	list($count) = $smcFunc['db_fetch_row']($request);
	$smcFunc['db_free_result']($request);
	return $count;
}

function UTM_Get_Topics($start, $items_per_page, $sort)
{
	global $smcFunc, $modSettings, $context, $user_info;

	$request = $smcFunc['db_query']('', '
		SELECT
			t.id_topic, t.num_replies, t.num_views, t.id_first_msg, m.id_msg_modified,
			' . ($user_info['is_guest'] ? '0' : 'COALESCE(lt.id_msg, COALESCE(lmr.id_msg, -1)) + 1') . ' AS new_from,
			m.id_member, COALESCE(mem.real_name, m.poster_name) AS poster, m.icon,
			m.subject AS subject, m.poster_time AS posted, b.id_board, b.name AS board_name
		FROM {db_prefix}topics AS t
			INNER JOIN {db_prefix}boards AS b ON (b.id_board = t.id_board)
			INNER JOIN {db_prefix}messages AS m ON (m.id_msg = t.id_first_msg)
			LEFT JOIN {db_prefix}members AS mem ON (mem.id_member = m.id_member)' . ($user_info['is_guest'] ? '' : '
			LEFT JOIN {db_prefix}log_topics AS lt ON (lt.id_topic = t.id_topic AND lt.id_member = {int:current_member})
			LEFT JOIN {db_prefix}log_mark_read AS lmr ON (lmr.id_board = t.id_board AND lmr.id_member = {int:current_member})') . '
		WHERE {query_see_board}' . (!empty($modSettings['recycle_enable']) && !empty($modSettings['recycle_board']) ? '
			AND b.id_board != {int:recycle_board}' : '') . (!empty($context['unanswered_member']) ? '
			AND m.id_member = {int:id_member}' : '') . '
			AND t.id_first_msg = t.id_last_msg
			AND t.locked = {int:not_locked}' . (!empty($context['days']) ? '
			AND m.poster_time > {int:time_limit}' : '') . (!empty($context['UAT_boards']) ? '
			AND b.id_board IN ({array_int:boards})' : '') . '
		ORDER BY {raw:sort}
		LIMIT {int:start}, {int:per_page}',
		array(
			'current_member' => $user_info['id'],
			'sort' => $sort,
			'start' => $start,
			'per_page' => $items_per_page,
			'not_locked' => 0,
			'time_limit' => time() - $context['days'] * 86400,
			'recycle_board' => $modSettings['recycle_board'],
			'id_member' => $context['unanswered_member'],
			'boards' => $context['UAT_boards'],
		)
	);
	$topics = array();
	while ($row = $smcFunc['db_fetch_assoc']($request))
		$topics[] = $row;
	$smcFunc['db_free_result']($request);
	return $topics;
}

/********************************************************************************
* Functions necessary to properly format the data for the display:
********************************************************************************/
function UTM_Icon(&$rowData)
{
	global $context, $settings;
	
	if (!isset($context['icon_sources'][$rowData['icon']]))
		$context['icon_sources'][$rowData['icon']] = file_exists($settings['theme_dir'] . '/images/post/' . $rowData['icon'] . '.gif') ? 'images_url' : 'default_images_url';
	return '<img src="' . $settings[$context['icon_sources'][$rowData['icon']]] . '/post/' . $rowData['icon'] . '.gif" alt="" />';
}

function UTM_Subject(&$rowData)
{
	global $scripturl, $txt, $settings, $forum_version;

	$board = '<strong><a href="' . $scripturl . '?board=' . $rowData['id_board'] . '.0">' . $rowData['board_name'] . '</a></strong>';
	$topic = '<strong><a href="' . $scripturl . '?topic=' . $rowData['id_topic'] . '.0">' . $rowData['subject'] . '</a></strong>';
	$user = '<strong><a href="' . $scripturl . '?action=profile;user=' . $rowData['id_member'] . '">' . $rowData['poster'] . '</a></strong>';
	if (substr($forum_version, 0, 7) == 'SMF 2.1')
		$unread = $rowData['new_from'] >= $rowData['id_msg_modified'] ? '' : ' <span class="new_posts">' . $txt['new'] . '</span>';
	else
		$unread = $rowData['new_from'] >= $rowData['id_msg_modified'] ? '' : ' <img src="' . $settings['lang_images_url'] . '/new.gif" alt="' . $txt['new'] . '" />';
	return $board . ' &raquo; ' . $topic . $unread . '<div class="smalltext">' . $txt['started_by'] . ' ' . $user . '</div>';
}

function UTM_Checkbox(&$rowData)
{
	global $context, $user_info;
	
	if (!isset($context['delete_own'][$rowData['id_board']]))
	{
		$own = $context['delete_own'][$rowData['id_board']] = allowedTo('delete_own', $rowData['id_board']);
		$any = $context['delete_any'][$rowData['id_board']] = allowedTo('delete_any', $rowData['id_board']);
	}
	else
	{
		$own = $context['delete_own'][$rowData['id_board']];
		$any = $context['delete_any'][$rowData['id_board']];
	}
	$context['can_delete'] |= ($delete = ($own && $rowData['id_member'] == $user_info['id']) || $any);
	if ($delete)
	{
		$_SESSION['unanswered_can_delete'][] = $rowData['id_topic'];
		return '<input type="checkbox" name="remove[' . $rowData['id_topic'] . ']" class="input_check" />';
	}
}

/********************************************************************************
* Our templates functions:
********************************************************************************/
function template_unanswered_topics()
{
	template_show_list('unanswered_topics');
}

function template_unanswered_cannot_delete()
{
}

?>